driver-core/src/main/com/mongodb/internal/operation/ClientBulkWriteOperation.java

```diff
@@ -180,23 +194,79 @@ public ClientBulkWriteResult execute(final WriteBinding binding) throws ClientBu
         return resultAccumulator.build(transformedTopLevelError, effectiveWriteConcern);
     }
 
+    @Override
+    public void executeAsync(final AsyncWriteBinding binding,
+                             final SingleResultCallback<ClientBulkWriteResult> finalCallback) {
+        WriteConcern effectiveWriteConcern = validateAndGetEffectiveWriteConcern(binding.getOperationContext().getSessionContext());
+        ResultAccumulator resultAccumulator = new ResultAccumulator();
+
+        beginAsync().<MongoException>thenSupply(callback -> {
+            executeAllBatchesAsync(effectiveWriteConcern, binding, resultAccumulator, (result, t) -> {
+                MongoException transformedTopLevelError = null;
+                if (t != null) {
+                    if (t instanceof MongoException) {
+                        transformedTopLevelError = transformWriteException((MongoException) t);
+                    } else {
+                        callback.completeExceptionally(t);
+                        return;
+                    }
+                }
+                callback.onResult(transformedTopLevelError, null);
+            });
+        }).<ClientBulkWriteResult>thenApply((transformedTopLevelError, callback1) -> {
+            callback1.onResult(resultAccumulator.build(transformedTopLevelError, effectiveWriteConcern), null);
+        }).finish(finalCallback);
+    }
+
     /**
      * To execute a batch means:
      * <ul>
      *     <li>execute a `bulkWrite` command, which creates a cursor;</li>
      *     <li>consume the cursor, which may involve executing `getMore` commands.</li>
      * </ul>
      *
-     * @throws MongoException When a {@linkplain ClientBulkWriteException#getError() top-level error} happens.
+     * @throws MongoException When a {@linkplain ClientBulkWriteException#getCause() top-level error} happens.
```

Comment:
Thank you! I missed this when I refactored out `ClientBulkWriteException.getError`.